lib/Monads/trace/Trace_RG.thy

```diff
@@ -230,6 +314,135 @@ lemma bind_twp[wp_split]:
+  by (fastforce simp: validIE_def validI_def split: sum.splits)
+
+lemma rg_post_imp_dc2E:
+  "\<lbrakk>\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> a \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_. Q'\<rbrace>; \<And>s0 s. Q' s0 s \<Longrightarrow> Q s0 s\<rbrakk> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> a \<lbrace>G\<rbrace>,-,\<lbrace>\<lambda>_. Q\<rbrace>"
+  by (fastforce simp: validIE_def validI_def split: sum.splits)
+
+lemma rg_guar_imp:
+  "\<lbrakk>\<And>s0 s. G' s0 s \<Longrightarrow> G s0 s; \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G'\<rbrace>,\<lbrace>Q\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  by (force simp: validI_def guar_cond_def)
+
+lemma rg_guar_impE:
+  "\<lbrakk>\<And>s0 s. G' s0 s \<Longrightarrow> G s0 s; \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G'\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  by (clarsimp simp: validIE_def elim!: rg_guar_imp)
+
+lemmas rg_strengthen_post = rg_post_imp[rotated]
+lemmas rg_strengthen_postE = rg_post_impE[rotated 2]
+lemmas rg_strengthen_guar = rg_guar_imp[rotated]
+lemmas rg_strengthen_guarE = rg_guar_impE[rotated]
+
+lemma rg_conjD1:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0. Q rv s0 and Q' rv s0\<rbrace> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. Q rv\<rbrace>"
+  unfolding validI_def by auto
+
+lemma rg_conjD2:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0. Q rv s0 and Q' rv s0\<rbrace> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. Q' rv\<rbrace>"
+  unfolding validI_def by auto
+
+lemma rg_post_disjI1:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. Q rv\<rbrace> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0. Q rv s0 or Q' rv s0\<rbrace>"
+  unfolding validI_def by auto
+
+lemma rg_post_disjI2:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. Q' rv\<rbrace> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0. Q rv s0 or Q' rv s0\<rbrace>"
+  unfolding validI_def by auto
+
+\<comment> \<open>FIXME MC: do we want the last assumption?  It might make unification easier in unusual cases but so
+           far is always just resolved with refl. \<close>
+lemma use_validI:
+  "\<lbrakk>(tr, Result (rv, s')) \<in> (rely f R s0 s); \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>; P s0 s; s0' = last_st_tr tr s0\<rbrakk>
+   \<Longrightarrow> Q rv s0' s'"
```

Comment:
Could always make a `'` version with the assumption in and then the main version where it is resolved by `refl`. Then it's available in the rarer cases where it's needed and we get the more convenient short version by default.

(nitpick: would leave out the parentheses in `(rely ..)`)